provider/plug_block_devices.py

```diff
@@ -0,0 +1,285 @@
+            images = images.split() if images else self.data_imgs
+            logging.info("Start to hotplug devices \"%s\" by monitor %s." % (
+                ' '.join(images), monitor.name))
+            self._create_block_devices(images)
+            self._plug_devices(HOTPLUG, self.hotplug_devs, monitor)
+            self.lock.release()
+        self._check_qmp_output(HOTPLUG)
+
+    def unplug_devices(self, images=None, monitor=None):
+        """
+        Unplug the devices which are defined by images.
+
+        :param images: Image tags, e.g, "stg0" or "stg0 stg1 stg2".
+        :type images: str
+        :param monitor: Qemu monitor object.
+        :type monitor: qemu_monitor.QMPMonitor
+        """
+        if self.lock.acquire():
+            monitor = monitor if monitor else self.vm.monitor
+            images = images.split() if images else self.data_imgs
+            if images is None:
+                self.unplug_devs = [dev for dev in self.hotplug_devs if not isinstance(
+                    dev, qdevices.QDrive)]
+            else:
+                self.unplug_devs = [dev for dev in self.vm.devices if dev.get_param(
+                    "id") in images]
+            logging.info("Start to unplug devices \"%s\" by monitor %s." % (
+                ' '.join(images), monitor.name))
+            self._plug_devices(UNPLUG, self.unplug_devs, monitor)
+            self.lock.release()
+        self._check_qmp_output(UNPLUG)
+
+
+class SerialPlugManger(PlugManager):
+    def __init__(self, vm, params):
+        super(SerialPlugManger, self).__init__(vm, params)
+
+    @PlugManager.check_devices_num(action=HOTPLUG)
+    def hotplug_devices(self, images=None, timeout=300):
+        """
+        Hot plug the devices which are defined by images.
+
+        :param images: Image tags, e.g, "stg0" or "stg0 stg1 stg2".
+        :type images: str
+        :param timeout: Timeout for hot plugging.
+        :type timeout: float
+        """
+        super(SerialPlugManger, self).hotplug_devices(images)
+
+    @PlugManager.check_devices_num(action=UNPLUG)
+    def unplug_devices(self, images=None, timeout=300):
+        """
+        Unplug the devices which are defined by images.
+
+        :param images: Image tags, e.g, "stg0" or "stg0 stg1 stg2".
+        :type images: str
+        :param timeout: Timeout for unplugging.
+        :type timeout: float
+        """
+        super(SerialPlugManger, self).unplug_devices(images)
+        self.check_deleted_event(timeout)
+
+
+class ParallelPlugManger(PlugManager):
+    def __init__(self, vm, params):
+        self._orig_disks = set()
+        super(ParallelPlugManger, self).__init__(vm, params)
+
+    def _parallel_plug_devices(self, action, images, timeout):
+        threads = []
+        n = 0
+        self.imgs = images.split() if images else self.data_imgs
+        self._orig_disks = self.list_all_disks()
+        for i, img in enumerate(self.imgs):
```

Comment:
there should be an upper bound of the thread count that might relate to the host CPU count.